src/main/java/org/folio/event/handler/HoldingUpdateAsyncRecordHandler.java

```diff
@@ -0,0 +1,193 @@
+import io.vertx.core.Future;
+import io.vertx.core.MultiMap;
+import io.vertx.core.Vertx;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.core.http.impl.headers.HeadersMultiMap;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.client.WebClient;
+import lombok.extern.log4j.Log4j2;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.tuple.Pair;
+import org.folio.cql2pgjson.exception.FieldException;
+import org.folio.event.dto.ResourceEvent;
+import org.folio.event.service.AuditOutboxService;
+import org.folio.okapi.common.GenericCompositeFuture;
+import org.folio.okapi.common.XOkapiHeaders;
+import org.folio.rest.jaxrs.model.Location;
+import org.folio.rest.jaxrs.model.OrderLineAuditEvent;
+import org.folio.rest.jaxrs.model.PoLine;
+import org.folio.rest.persist.Conn;
+import org.folio.rest.persist.DBClient;
+import org.folio.services.lines.PoLinesService;
+import org.folio.spring.SpringContextUtil;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.folio.event.InventoryEventType.INVENTORY_HOLDING_UPDATE;
+
+@Log4j2
+public class HoldingUpdateAsyncRecordHandler extends InventoryUpdateAsyncRecordHandler {
+
+  public static final String ID = "id";
+  public static final String INSTANCE_ID = "instanceId";
+  public static final String PERMANENT_LOCATION_ID = "permanentLocationId";
+  public static final String PO_LINE_LOCATIONS_HOLDING_ID_CQL = "locations==*%s*";
+  public static final String STORAGE_HOLDING_URL = "%s/holdings-storage/holdings/%s";
+
+  @Autowired
+  private PoLinesService poLinesService;
+
+  @Autowired
+  private AuditOutboxService auditOutboxService;
+
+  private final WebClient webClient;
+
+  public HoldingUpdateAsyncRecordHandler(Vertx vertx, Context context) {
+    super(INVENTORY_HOLDING_UPDATE, vertx, context);
+    webClient = WebClient.create(vertx);
+    SpringContextUtil.autowireDependencies(this, context);
+  }
+
+  @Override
+  protected Future<Void> processInventoryUpdateEvent(ResourceEvent resourceEvent, Map<String, String> headers,
+                                                     String tenantId, DBClient dbClient) {
+    var holdingObject = JsonObject.mapFrom(resourceEvent.getNewValue());
+    var holdingId = holdingObject.getString(ID);
+    return dbClient.getPgClient()
+      .withTrans(conn -> {
+        try {
+          return processPoLinesUpdate(resourceEvent, headers, tenantId, holdingId, conn)
+            .compose(poLines -> {
+              var adjacentHoldingIds = extractDistinctAdjacentHoldingsToUpdate(holdingId, poLines);
+              return updateAdjacentHoldingsWithNewInstanceId(resourceEvent, headers, adjacentHoldingIds);
+            });
+        } catch (FieldException e) {
+          throw new IllegalStateException("Produced and invalid CQL wrapper", e);
```

Comment:
do you think we need this catch? there is can be sitatuations when we could not control cql wrappers and want to logs exception about this with rethrowing?

Reply:
> do you think we need this catch? there is can be sitatuations when we could not control cql wrappers and want to logs exception about this with rethrowing?

Yea its not needed, I will make it a runtime exception so we don't need to handle or rethrow it

Reply:
if it checked exception we also can consider lombock to do the same to simplify code

Reply:
> if it checked exception we also can consider lombock to do the same to simplify code

`@SneakyThrows`?